### PR TITLE
shared: add edit-util (part of which extracted from systemctl-edit)

### PR DESCRIPTION
**Pull request by YHNdnzj**
I'm recently working on #26246, and planning to make this new shared util a prerequisite for that :)

### CODE REVIEW

**Comment by amit-tewari** [comment]
@YHNdnzj 
Nice work on splitting the util to it's own files.
https://github.com/systemd/systemd/pull/26481 has an idea about how this may be implemented, without needing another option, or changing behavior of existing options. But that approach may be redundant with your implementation, which seems a far reaching change on how edit works. In case you think I should work in parallel, please let me know or I can close my PR.

**Comment by YHNdnzj** [comment]
@amit-tewari I prefer adding a new option because I think `--full` doesn't carry a (clear) meaning of creating a new unit, but I am not really against it either. Opinions from others are welcome :)

> In case you think I should work in parallel, please let me know or I can close my PR.

Well, I'm hoping that this PR can be reviewed soon. If you don't mind, maybe we can delay your PR until this gets merged? I won't ask for closing tho, since we haven't made a decision on which way to go. Appreciate your help :D

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c:150]: Previously, this was `else if`. Is this intended?

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: The second `comment_paths` check is redundant.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c:181]: In the assertion at the beginning of the function, `marker_start` _or_ `marker_end` is requested. Hence, one of them can be NULL. Is it intended? If so, please do not print `(null)`.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: It is not necessary to set the flag earlier. Please check error earlier.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: When this is triggered, `edit_file.path` is leaked. I do not think it is necessary to introduce a temporal variable `edit_file`, but simply assign `(*ef)[n].path` or so.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: I'd like to have `edit_file_free_and_unlink_all()`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: BTW, if you clean up the code, then I'd like to have
```c
struct EditFiles {
        EditFile *files;
        size_t n_files;
};
```
Then, we can drop `edit_files_count()`, and the count can be simply used in `GREEDY_REALLOC()`.
If you feel the struct name is confusing, then `EditFileContext` may be also good.

**Review by yuwata — COMMENTED** [review]
Unlike introducing similar but cleaned-up code in edit-util.[ch], first extract code from systemctl-edit.c as is, and then apply the cleaning-ups. Otherwise, it is hard to review...

File: src/shared/edit-util.c
Comment (YHNdnzj) [reply src/shared/edit-util.c:150]: Yes. I think the two don't really conflict, and can be used together if necessary.

File: src/shared/edit-util.c
Comment (YHNdnzj) [reply src/shared/edit-util.c]: Hmm, I'm not very sure who should be responsible for `unlink`-ing. Should it be the caller using a cleanup attribute or we call the cleanup function here?

File: src/shared/edit-util.c
Comment (yuwata) [reply src/shared/edit-util.c:150]: Maybe so, but I am not sure. Please do that change in a separate commit with explanation.

File: src/shared/edit-util.c
Comment (yuwata) [reply src/shared/edit-util.c]: I do not think so. If possible (and in this case it _is_ possible, I think), functions should not take any side effect on failure. So, please remove temporary files on failure.

**Comment by YHNdnzj** [comment]
Converting this to draft since there're still much more things to do 😉

**Comment by YHNdnzj** [comment]
@yuwata Split into multiple commits and addressed all the comments. PTAL :)

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Hm? Maybe,
```
assert(!comment_paths || (marker_start && marker_end));
```

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: If the above comment is correct, then `strempty()` can be dropped, though not a big matter.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Also here, `strempty()` can be dropped.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: The condition about the line number is dropped. Is it intended? If so, why?

File: src/shared/edit-util.h
Comment (yuwata) [inline src/shared/edit-util.h]: Maybe better to rewrite with EditFileContext.n_files ??

File: src/shared/edit-util.h
Comment (yuwata) [inline src/shared/edit-util.h]: I'd like to do such by a function, instead of macro.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c:223]: Oh, please do not drop these empty lines.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c]: Please check by `context.n_files == 0`. The buffer may be allocated, but may not be used.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Should be merged to the third commit?

**Review by yuwata — CHANGES_REQUESTED** [review]
Sorry for late. Only the first 5 commits are reviewed.

File: src/shared/edit-util.h
Comment (YHNdnzj) [reply src/shared/edit-util.h]: I don't know exactly how since one can't initialize variables with different types in a for loop?

**Comment by yuwata** [comment]
CIs are red...

File: src/shared/edit-util.h
Comment (YHNdnzj) [reply src/shared/edit-util.h]: > CIs are red...

Because of this...

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [reply src/systemctl/systemctl-edit.c:223]: The comment is for the 4th commit. You unnecessarily removed empty lines in the 4th commit, and re-add some of them in a later commit...

Comment (YHNdnzj) [reply src/systemctl/systemctl-edit.c:223]: Oops, sorry

File: src/shared/edit-util.h
Comment (yuwata) [inline src/shared/edit-util.h]: I mean something like the following:
```c
for (EditFile *s = c.files; s < c.files + c.n_files; s++)
```
But, after reconsidering about the macro, now I do not think it is necessary to introduce the macro. The pattern is very common (e.g. see systemctl-list-units.c) and not hard to read. (Of course, the original macro conditioned with the existence of `EditFile.path` may not be trivial, but after `EditFileContext` being introduced, the loop is now quite simple.)

Comment (YHNdnzj) [reply src/shared/edit-util.h]: It still feels tidier so I'd like to preserve it 🤔

Comment (yuwata) [reply src/shared/edit-util.h]: OK. The opinion about dropping the macro is not strong opinion. Feel free to take any form you'd like.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c]: Still, empty lines are dropped unnecessarily in the 4th commit...

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c:649]: This line is dropped in the new `do_edit_files_and_install()` function.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c]: Why you dropped the error handling?? Please keep it.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c]: Maybe, better to include this in `edit_context_done()`?

File: src/shared/edit-util.h
Comment (yuwata) [inline src/shared/edit-util.h]: Hmm... This evaluates `c` multiple times... I really do not like the macro.
At least, please make the macro takes `EditFileContext*` rather than `EditFileContext`.

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c]: Hm? Why the change is necessary? When we reach here, then there is no error we should propagate. Right?

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Hm? Why not to use `args = TAKE_PTR(editor_args);`??

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Please use `DECIMAL_STR_MAX`:
```
char l[1+DECIMAL_STR_MAX];
```

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: You can use `strv_extendf()`

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: You can merge them.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: `strv_prepend()` internally calls `strdup()`, hence the above `strdup()` is not necessary.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: You can use `free_and_strdup()`.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Please move this to the child block.

I think it is better to move the child operation into a separated function:
```c
r = safe_fork(...);
if (r < 0)
        return r;
if (r == 0) {
        r = run_editor_child(...);
        _exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
}
return 0;
```

Then, you can shorten the error handling in `run_editor_child()` something like the following:
```c
r = strv_extend(...);
if (r < 0)
        return log_oom();
```

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/systemctl/systemctl-edit.c
Comment (YHNdnzj) [reply src/systemctl/systemctl-edit.c]: Hmm, I think it's actually a rare case for us to remove the parent directory. Maybe add this to a separate function?

Comment (yuwata) [reply src/systemctl/systemctl-edit.c]: > Hmm, I think it's actually a rare case for us to remove the parent directory.

Not sure, but the loop should be run on finalization (even on the failure cases), IIUC. So, I think this should be called in the cleanup function.

Comment (YHNdnzj) [reply src/systemctl/systemctl-edit.c]: >  Not sure, but the loop should be run on finalization (even on the failure cases)

That's why it's in the `end` block?

But the actual probleam is that cleanup functions doesn't take extra parameters. But I think we can add a switch to `EditFileContext` for this.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: `original unit file` -> `original file` ??

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Failure in rebase.

File: src/shared/edit-util.c
Comment (yuwata) [reply src/shared/edit-util.c]: Not addressed in the 4th commit?

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Oh...

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Nah, in the 4th commit, `args` is allocated with `newa()`. Hence, we cannot use `strv_extend()`.

**Review by yuwata — CHANGES_REQUESTED** [review]
The 4th commit is broken...

File: src/shared/edit-util.c
Comment (YHNdnzj) [reply src/shared/edit-util.c]: I should have stayed on my bed...

File: src/shared/edit-util.c
Comment (YHNdnzj) [reply src/shared/edit-util.c]: I think it does strdup internally? But I'm not familiar with `alloca`...

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Please declare the special variable `r` at the beginning of the function.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Now the error will be ignored. Please downgrade the log level (maybe `log_debug_errno()`), and append `", ignoring"` in the log message.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c:49]: Not a big matter, but we usually do not assign `NULL` or `0` in `xyz_done()`. But _done_ in `xyz_clear()`.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Please add a period at the end of the log message.

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: This is not necessary, as the temporary files and unnecessary directories will be removed in `edit_file_context_done()` called in `verb_edit()`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks much better now.

**Comment by YHNdnzj** [comment]
`edit_files_contains` is causing undefined behaviors in `EDIT_FILES_FOREACH`. I'll just wait for #26678 to be merged and rebase this to use it instead :)

**Comment by YHNdnzj** [comment]
> edit_files_contains is causing undefined behaviors in EDIT_FILES_FOREACH. I'll just wait for https://github.com/systemd/systemd/pull/26678 to be merged and rebase this to use it instead :)

Done 😉

**Comment by YHNdnzj** [comment]
@yuwata It appears that `num` needs to be converted to non-const in `_FOREACH_ARRAY`...

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Why we need these two?? Please free the array unconditionally, and always use `_done()` in the cleanup attribute, instead of `_clear()`

File: src/shared/edit-util.c
Comment (yuwata) [inline src/shared/edit-util.c]: Not a matter, but `const` qualifier can be added for `context` ?

File: src/systemctl/systemctl-edit.c
Comment (yuwata) [inline src/systemctl/systemctl-edit.c:640]: I'd like to keep the comment in the new function `do_edit_files_and_install()`

**Review by yuwata — CHANGES_REQUESTED** [review]
Looks pretty better now.

**Review by yuwata — APPROVED** [review]
LGTM. Thank you!

**Comment by YHNdnzj** [comment]
Updated commit messages to match the latest changes. Thanks to yuwata for the patient review :)

File: src/shared/edit-util.c
Comment (poettering) [inline src/shared/edit-util.c:30]: i->temp = unlink_and_free(i->temp)

File: src/shared/edit-util.c
Comment (poettering) [inline src/shared/edit-util.c:40]: we usually add debug logs on unexpected errors. i.e. if we see something else than ENOENT/ENOTEMPTY we should probably at least debug log

Comment (poettering) [reply src/shared/edit-util.c:40]: also, parent can be NULL here. please check that first.

File: src/shared/edit-util.c
Comment (poettering) [inline src/shared/edit-util.c:57]: path_equal()

**Review by poettering — COMMENTED** [review]
sorry for the late review, but some minor things

File: src/shared/edit-util.c
Comment (YHNdnzj) [reply src/shared/edit-util.c:30]: This has been applied in https://github.com/systemd/systemd/pull/26776 :)

**Pull request merged**
